polyply/src/nanoparticle_constructor.py

```diff
@@ -0,0 +1,354 @@
+        self.graph = MetaMolecule._block_graph_to_res_graph(
+            self.np_molecule
+        )  # generate residue graph
+
+        # generate meta molecule from the residue graph with a new molecule name
+        self.meta_mol = MetaMolecule(
+            self.graph, force_field=self.force_field, mol_name="test"
+        )
+
+        self.np_molecule.meta["moltype"] = "TEST"
+        # reassign the molecule with the np_molecule we have defined new interactions with
+        self.meta_mol.molecule = self.np_molecule
+        return self.meta_mol
+
+    def create_itp(self, write_path):
+        """
+        create itp of the final concatenated ligand-functionalized nanoparticle file
+        """
+        with open(str(write_path), "w") as outfile:
+            write_molecule_itp(self.np_molecule, outfile)
+
+    def create_gro(self, write_path):
+        """
+        create gro file from the system
+        """
+        self.top.box = np.array([3.0, 3.0, 3.0])
+        system = self.top.convert_to_vermouth_system()
+        vermouth.gmx.gro.write_gro(
+            system,
+            write_path,
+            precision=7,
+            title="these are random positions",
+            box=self.top.box,
+            defer_writing=False,
+        )
+
+
+def gen_np(
+    library,
+    out_path,
+    inpath,
+    core_name,
+    ligand_names,
+    ligand_anchor,
+    pattern="random",
+    name="nanoparticle",
+    box=np.array([10.0, 10.0, 10]),
+):
+    """
+
+    function that can create a flexible nanoparticle
+
+    How about we define a function here that is called gen_np analogously
+    to the gen_coords or gen_params. This function should have the
+    arguments that are tunable to build the NP.
+
+    """
+    # 1. load library as described in the second comment
+    force_field = load_library(name, library, inpath)
+    # 2. select the core block and generate coordinates of the core
+    core_mol = force_field.blocks[core_name].to_molecule()
+    core_mol = NanoparticleCoordinates().run_molecule(
+        core_mol
+    )  # not entirely certain what this part does..
+
+    # 3. initiate a class NPAttachLigandsGraphs, that generates only the itp files
+    full_np_metamol = NanoparticleGenerator(
+        force_field, ligand_names, 1, ligand_anchor, None
+    ).run(core_mol)
+    # probably a good idea to let this function return a meta_molecule
+    # once we have the metamolecule we can instantiate a topology object to use later
+
+    np_top = Topology(name=name, force_field=force_field)
+    np_top.molecules = [full_np_metamol]
+    # 4. Generate coordinates
+    # 5. Write output like you've done before
+    # first the itp file
+    with open(str(out_path) + "/" + "NP.itp", "w") as outfile:
+        write_molecule_itp(np_top.molecules[0].molecule, outfile)
+
+    # then the coordinate file
+    # command = " ".join(sys.argv)
+    system = np_top.convert_to_vermouth_system()
+    NanoparticleLigandCoordinates().run_system(system)
```

Comment:
instead of this, you can provide the `np_top` to generate templates like so:

```
# this line is needed in order for it to skip generation of the core coordinates. 
full_np_metamol.templates = {"rename of core": core coordinates}
LigandGenerator = GenerateTemplates(topology=np_top, max_opt=10).run_system(np_top)
```

This will generate all Ligand coordinates, which can be accessed through `full_np_metamol.templates`